### ISSUE TITLE
Extension doesn't just work out of the box on Windows with RBAC enabled

### ISSUE THREAD

**Issue by davidmatson** [Microsoft/ApplicationInsights-Kubernetes#171]
I wasn't able to use the troubleshooting instructions (see #170), but I manually tried making an HTTPS request to api/v1/namespaces/default/pods, and that fails with the following error:
```
Response: Forbidden
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods is forbidden: User \"system:servicea
ccount:default:default\" cannot list resource \"pods\" in API group \"\" in the namespace \"default\"","reason":"Forbidd
en","details":{"kind":"pods"},"code":403}
```

Here's my test:
```c#
using System;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;

static class Program
{
    public static void Main()
    {

        string kubernetesServiceHost = Environment.GetEnvironmentVariable(@"KUBERNETES_SERVICE_HOST");

        string kubernetesServicePort = Environment.GetEnvironmentVariable(@"KUBERNETES_SERVICE_PORT");

        string baseAddress = string.Format("https://{0}:{1}/", kubernetesServiceHost, kubernetesServicePort);

        string token = File.ReadAllText(@"C:\var\run\secrets\kubernetes.io\serviceaccount\token");

        ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;

        using (HttpClientHandler handler = new HttpClientHandler { ServerCertificateCustomValidationCallback = (c, d, e, f) => true })
        using (HttpClient client = new HttpClient(handler))
        using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, baseAddress + "api/v1/namespaces/default/pods"))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

            using (HttpResponseMessage response = client.SendAsync(request).GetAwaiter().GetResult())
            {
                Console.WriteLine("Response: {0}", response.StatusCode);
                string content = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                Console.WriteLine(content);
            }
        }
    }
}
```

Even if that step worked, I don't think this extension would work on this node though, at least from a quick scan through the source of GetMyPodAsync, because it wouldn't know which pod is mine, because the environment variable APPINSIGHTS_KUBERNETES_POD_NAME is not defined, and the only other mechanism is via Container ID, which won't work either according to this source comment:
```c#
// Container id won't be fetched for windows container.
```

Is this extension expected to work on a general Windows agent node?

**Comment by davidmatson** [comment]
I think the first problem is related to RBAC - the default service account (which is the token given to the pod) doesn't have access to list all pods. I can manually create a clusterrolebinding to give the default pod service account this access, but that isn't the default.

However, even if I were to do so, that still wouldn't solve the problem of the pod figuring out which one is its own entry.

**Comment by xiaomi7732** [comment]
@davidmatson Please follow the example here and try it again: https://github.com/Microsoft/ApplicationInsights-Kubernetes/tree/develop/examples/WindowsContainer
Feel free to open this if that doesn't address your issue.

**Comment by davidmatson** [comment]
I'm not using ASP.NET, so that example doesn't entirely apply. Also, I don't believe it addresses the RBAC issue raised. I see the documentation about APPINSIGHTS_KUBERNETES_POD_NAME, but that only resolves one of the two problems I encountered.
Could you re-open this issue for me? I apparently don't have permissions to do so.

**Comment by xiaomi7732** [comment]
Hey @davidmatson 
1. I thought you figured the RBAC part already, right? Otherwise, please reference [RBAC section](https://github.com/Microsoft/ApplicationInsights-Kubernetes/tree/develop/examples/BasicUsage_clr21_RBAC#setup-the-default-service-account-for-rbac-enabled-cluster).

2. Downward API should just work for anything managed by K8s. As long as it is set in the yaml, the value will be there. Once set, read it from the environment variable of the container. Does that address the issue?

**Comment by davidmatson** [comment]
Thanks - I hadn't found that RBAC section before, that explains that part.

Is there any way to have the extension work without having to do a custom service account on every pod? Elsewhere, app insights "just works" as long as the app has the instrumentation key and outbound internet access. (I tended to think of the simplicity of app insights setup as one of its main advantages compared to other options like log analytics.)

**Comment by davidmatson** [comment]
Yeah, I guess I was hoping for a way to add Kubernetes info which would just work as well (for example, the other web jobs extension didn't require any extra configuration to get it working).

When RBAC is enabled, is there any information we can populate that would "just work" if no custom service principal is specified? On linux, we'd have the pod name - is that it?

**Comment by xiaomi7732** [comment]
@davidmatson I am sorry that I think there are some confusions. RBAC and the pod name are 2 different issues and they do NOT relate to each other.

The Kubernetes info appends to the telemetry are all queried from the master using the service account. When RBAC is not enabled, the sa by default having permission to query the master and it works; When RBAC is enabled, the sa do not have permission to query the master, thus, needs the bindings;

Querying the current pod, on the other hand, is a different issue. Let's say when there's no RBAC, even on Linux, there's no API returning the current Pod. The way it works is we cross-reference the container id for the pod and the container id from the cgroup to identify the pod. On Windows, though, there's no cgroup, that we lose the handle to the container id, which is why we can't determine which pod the code is running within. The way works are that the pod name is passed in through the downward api.

All in all, to make things work per current implementation, sa must have permission to query the master; additionally, on Windows, the environment variable is also needed to be set.

**Comment by davidmatson** [comment]
Yes, I understand the difference in these two issues. My focus here is on the RBAC side of things.

On linux, when RBAC is enabled, what pod-related information is available from the container without needing a custom service account/permissions? Just the container ID?

**Comment by xiaomi7732** [comment]
@davidmatson That is a good question.

Basically, what you are asking is: without querying master (since no permission), what info can I get out of a container managed by K8s.

I am afraid that is out of my expertise.

**Comment by davidmatson** [comment]
Sure, makes sense.

Last remaining question for this issue then - could we update the documentation to make the service account prerequisite more prominent? For example, there's no mention of it on this page that I could find:
https://github.com/Microsoft/ApplicationInsights-Kubernetes/wiki/Getting-Started-for-.NET-Core-Applications

(same applies for windows container name prerequisite)

**Comment by xiaomi7732** [comment]
@davidmatson Sure. That makes sense. I'll keep this open for tracking.
